File: java/java-samples/src/main/java/tech/techstay/generic/GenericClass.java
package tech.techstay.generic;

public class GenericClass {
  public static void main(String[] args) {
    MyValue<Integer> v1 = new MyValue<>(42);
    System.out.println(v1.getValue());

    MyValue<Double> v2 = new MyValue<>(3.14);
    System.out.println(v2.getValue());

  }
}


class MyValue<T> {
  private final T value;

  MyValue(T v) {
    value = v;
  }

  T getValue() {
    return value;
  }
}

File: scala/scala-example/src/main/scala/tech/techstay/typing/GenericClass.scala
package tech.techstay.typing

class Stack[A]:
  private var elements: List[A] = Nil

  def push(x: A): Unit =
    elements = x :: elements

  def peek: A = elements.head

  def pop(): A =
    val head = peek
    elements = elements.tail
    head

@main def genericClass(): Unit =
  val stack = Stack[Int]
  stack.push(1)
  stack.push(2)
  stack.push(3)

  println(stack.peek)
  println(stack.pop())
  println(stack.pop())
  println(stack.pop())
